yt_dlp/extractor/plex.py

```diff
@@ -0,0 +1,464 @@
+import urllib.parse
+
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    determine_ext,
+    float_or_none,
+    int_or_none,
+    parse_age_limit,
+    traverse_obj,
+)
+
+
+class PlexWatchBaseIE(InfoExtractor):
+    _NETRC_MACHINE = 'plex'
+
+    # Obtained from https://plex.tv/media/providers?X-Plex-Token=<plex_token>
+    _CDN_ENDPOINT = {
+        'vod': 'https://vod.provider.plex.tv',
+        'live': 'https://epg.provider.plex.tv',
+        'tv.plex.provider.epg': 'https://epg.provider.plex.tv',
+        'tv.plex.provider.vod': 'https://vod.provider.plex.tv',
+        'tv.plex.provider.metadata': 'https://metadata.provider.plex.tv',
+    }
+
+    _TOKEN = None
+    _CLIENT_IDENTIFIER = None
+
+    def _handle_login_error(self, error, error_message='', fatal=True):
+        error_json_message = self._parse_json(error.cause.read(), 'login_error')['errors'][0]['message']
+        if error.cause.code == 429:
+            # only non-fatal at error 429
+            if fatal:
+                raise ExtractorError(f'{error_json_message} {error_message}', cause=error.cause)
+            else:
+                self.report_warning(f'There\'s error on login : {error_json_message}, caused by {error.cause} {error_message}')
+        else:
+            raise ExtractorError(f'{error_json_message} {error_message}', cause=error.cause)
+
+    def _initialize_pre_login(self):
+        # TO DO: find better way to get cookie
+        # request to random page in plex.tv to get clientIdentifier in cookie
+        if PlexWatchBaseIE._TOKEN is None:
+            client_id = self._request_webpage(  # noqa: F841
+                'https://watch.plex.tv/', 'client_id', note='Downloading html page to get clientIdentifier')
+            cookie_ = {cookie.name: cookie.value for cookie in self.cookiejar}
+            self._CLIENT_IDENTIFIER = cookie_.get('clientIdentifier')
+
+    def _perform_login(self, username, password):
+        try:
+            resp_api = self._download_json(
+                'https://plex.tv/api/v2/users/signin', None, note='Logging in',
+                query={'X-Plex-Client-Identifier': self._CLIENT_IDENTIFIER},
+                data=f'login={username}&password={password}&rememberMe=true'.encode(),
+                headers={'Accept': 'application/json'}, expected_status=429)
+            PlexWatchBaseIE._TOKEN = resp_api.get('authToken')
+        except ExtractorError as e:
+            self._handle_login_error(e, fatal=False)
```

Comment:
It can't currently handle all ExtractorError. Either make it handle other cases, or re-raise them directly here. Like this:

```suggestion
            if not isinstance(e.cause, urllib.error.HTTPError):
                raise
            self._handle_login_error(e, fatal=False)
```